tests/testthat/test-parents.R

```diff
@@ -0,0 +1,247 @@
+
+  expected_jk <- join_keys(
+    join_key("df1", "df1", c("id", "id2")),
+    join_key("df1", "df2", c("id" = "id")),
+    join_key("df1", "df3", c("id" = "id")),
+    join_key("df2", "df2", c("id", "id2")),
+    join_key("df2", "df3", c("id", "id2")),
+    join_key("df3", "df3", c("id", "id2"))
+  )
+  parents(expected_jk) <- list(df1 = character(0), df2 = "df1", df3 = "df1")
+  testthat::expect_equal(jk, expected_jk)
+})
+
+# -----------------------------------------------------------------------------
+#
+# assert_parent_child
+
+test_that("assert_parent_child will detect empty keys", {
+  jk <- join_keys()
+  jk[["ds1"]][["ds2"]] <- character(0)
+  parents(jk) <- list(ds1 = "ds2")
+  expect_error(assert_parent_child(jk))
+})
+
+test_that("assert_parent_child will detect invalid key pairs", {
+  jk <- join_keys()
+  jk[["ds1"]][["ds2"]] <- "key1"
+  jk[["ds2"]][["ds1"]] <- character(0)
+  parents(jk) <- list(ds1 = "ds2")
+  expect_error(assert_parent_child(jk))
+
+  jk2 <- join_keys()
+  jk2[["ds2"]][["ds1"]] <- "key1"
+  jk2[["ds1"]][["ds2"]] <- character(0)
+  parents(jk2) <- list(ds1 = "ds2")
+  expect_error(assert_parent_child(jk2))
+})
+
+test_that("assert_parent_child will skip empty join_keys", {
+  jk <- join_keys()
+  expect_silent(assert_parent_child(jk))
+})
+
+testthat::test_that("assert_parent_child does nothing if no parents are present", {
+  jk <- join_keys()
+  join_keys(jk) <- list(join_key("df1", "df1", c("id" = "id")))
+  testthat::expect_identical(parents(jk), list())
+  testthat::expect_silent(assert_parent_child(jk))
+})
+
+testthat::test_that("assert_parent_child throws error if no join_keys exist for child-parent", {
+  jk <- join_keys()
+  join_keys(jk) <- list(join_key("df1", "df1", c("id" = "id")))
+  parents(jk) <- list(df1 = character(0), df2 = "df1", df3 = "df1")
+  testthat::expect_error(
+    assert_parent_child(jk),
+    "No join keys from df2 to its parent \\(df1\\) and vice versa"
+  )
+})
+
+testthat::test_that("assert_parent_child throws error if no join_keys exist for child-parent", {
+  jk <- join_keys()
+  join_keys(jk) <- list(
+    join_key("df1", "df1", c("id" = "id"))
+  )
+  # Change class as trick to allow for corrupt join_keys
+  class(jk) <- "list"
+  jk[["df2"]][["df1"]] <- "id"
+  class(jk) <- class(new_join_keys())
+  parents(jk) <- list(df1 = character(0), df2 = "df1", df3 = "df1")
+  testthat::expect_error(
+    assert_parent_child(jk),
+    "No join keys from df2 parent name \\(df1\\) to df2"
+  )
+})
+
+testthat::test_that("assert_parent_child throws error if no join_keys exist for child-parent", {
+  jk <- join_keys()
+  join_keys(jk) <- list(
+    join_key("df1", "df1", c("id" = "id"))
+  )
+  class(jk) <- "list"
+  jk[["df1"]][["df2"]] <- "id"
+  class(jk) <- class(new_join_keys())
+  parents(jk) <- list(df1 = character(0), df2 = "df1", df3 = "df1")
+  testthat::expect_error(
+    assert_parent_child(jk),
+    "No join keys from df2 to its parent \\(df1\\)"
+  )
+})
```

Comment:
Done, tested when setting parents 033bb497

Reply:
This change adds a validation when setting `parents` of `join_keys`, i.e. it will check if there is a minimal set of foreign keys that allow the parent setting.

What do you guys think of this behavior?